<launch>

	<!-- This file launches:
	 	* Stage
	 	* nav2d
	 		-Operator
	 		-Navigator
	 		-GetMap
	 		-Explore
	 		-SetGoal
	 		-joystick
	 		-Remote
	 		-exploration
	 	* rosbag record to record the /base_pose_ground_truth topic in a bag file
	 	* tf listener to write the gmapping tf
	 	* gmapping
	 	* RVIZ
	 	Params:
	 		-worldfile: the world file for stage
	 		-bag: the name of the bag file to write the ground truth from stage
	 		-output: the name of the file to write the gmapping log
	 	 -->

	<!-- Some general parameters -->
	<param name="use_sim_time" value="true" />

	<!-- Start Stage simulator with a given environment -->
	<node name="Stage" pkg="stage_ros" type="stageros" args="-g /$(arg worldfile)">
		<param name="base_watchdog_timeout" value="0" />
	</node>

	<!-- Record ground truth values -->
	<node name="recorder" pkg="rosbag" type="record" args="--lz4 -O $(arg bag) /base_pose_ground_truth /odom /base_scan"/>

	<!-- Write log odom gmapping for metricevaluator -->
	<node name="listener" pkg="tf_listener" type="tf_listener.py" args="$(arg outputfile)" />

	<!-- Start exploration of ambient -->

	<!-- Start Mapper to generate map from laser scans -->
	<node name="Mapper" pkg="gmapping" type="slam_gmapping" args="_xmin:=-100 _ymin:=-100 _ymax:=100
     _xmax:=100 _ogain:=3.0 _minimumScore:=0 _angularUpdate:=0.25 _linearUpdate:=1.0
     _map_update_interval:=5.0 _temporal_update:=5.0 _odom_frame:=odom _particles:=40 scan:=base_scan
     _srr:=0.01 _srt:=0.02 _str:=0.01 _stt:=0.02 _delta:=0.05 _maxUrange:=30 _maxRange:=60" />
	<arg name="robot_name" default="turtlebot" />
    <arg name="robot_prefix" default=""/>
    <arg name="log_path" />
    <arg name="output" default="screen"/>
    <arg name="frontier_selection" default="5"/>
    <arg name="use_sim_time" default="true" />
    <param name="use_sim_time"  value="$(arg use_sim_time)"/>


	<node pkg="move_base" type="move_base" respawn="true" name="move_base" output="log" clear_params="true">
<rosparam file="$(find explorer)/param/pioneer3at_costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find explorer)/param/pioneer3at_costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find explorer)/param/pioneer3at_local_costmap_params.yaml" command="load" />
    <rosparam file="$(find explorer)/param/pioneer3at_global_costmap_params.yaml" command="load" />
    <rosparam file="$(find explorer)/param/base_local_planner_params.yaml" command="load" />
    <param name="base_global_planner" value="global_replanner/GlobalReplanner"/>
    <param name="base_local_planner" value="base_local_planner/TrajectoryPlannerROS"/>
	</node>


    <!-- SIMULATION -->
    <node pkg="explorer" type="explorer" name="explorer" output="$(arg output)" respawn="true" clear_params="true">
	<!-- costmaps -->
	<param name="local_costmap/robot_base_frame" value="base_link" />
    <param name="local_costmap/global_frame" value="odom" />
    <param name="robot_prefix" value="$(arg robot_prefix)"/>
	<param name="global_costmap/robot_base_frame" value="base_link" />
	<param name="global_costmap/global_frame" value="map" />
	<param name="global_costmap/map_topic" value="/map" />
	<param name="local_costmap/map_topic" value="/map" />
	<param name="local_costmap/observation_sources" value="base_scan" />
	<param name="local_costmap/base_scan/topic" value="/base_scan" />
	<param name="local_costmap/base_scan/data_type" value="LaserScan" />
	<param name="local_costmap/base_scan/marking" value="true" />
	<param name="local_costmap/base_scan/clearing" value="true" />
	<param name="local_costmap/base_scan/sensor_frame" value="base_laser_link"/>
	<param name="observation_persistence" value="0.1" />

	<!-- simulation parameters -->
        <param name="log_path" value="$(arg log_path)" />
	<param name="robot_name" value="$(arg robot_name)" />

	<!-- move_base -->
	<param name="base_link" value="/base_link" />
	<param name="move_base_frame" value="/map" />


	<!-- exploration parameters -->
	<param name="frontier_selection" value="$(arg frontier_selection)" />
	<rosparam file="$(find explorer)/param/pioneer3at_costmap_common_params.yaml" command="load" ns="global_costmap" />
	<rosparam file="$(find explorer)/param/pioneer3at_costmap_common_params.yaml" command="load" ns="local_costmap" />
	<rosparam file="$(find explorer)/param/pioneer3at_local_costmap_params.yaml" command="load" />
	<rosparam file="$(find explorer)/param/pioneer3at_global_costmap_params.yaml" command="load" />

    	<remap from="/$(arg robot_name)/$(arg robot_name)/map" to="/$(arg robot_name)/map" />
    	<!-- <remap from="/$(arg robot_name)/$(arg robot_name)/base_scan" to="/$(arg robot_name)/base_scan" /> -->
    </node>

  <rosparam command="load" file="$(find range_proximity_safety_controller)/config/range_proximity_safety_ctrl_param" />

  <node name="range_proximity_safety_controller" type="range_proximity_safety_controller_node" pkg="range_proximity_safety_controller" output="screen" clear_params="true" >
    <remap from="scan" to="base_scan" />
    <remap from="cmd_vel" to="cmd_vel" />
  </node>

</launch>
